crates/core/tedge_api/src/entity_store.rs

```diff
@@ -0,0 +1,539 @@
+            external_id_index: HashMap::from([(external_id, main_device.mqtt_id)]),
+        })
+    }
+
+    /// Returns information about an entity under a given MQTT entity topic.
+    pub fn get(&self, entity_id: &str) -> Option<&EntityMetadata> {
+        self.entities.get(entity_id)
+    }
+
+    /// Returns information for an entity under a given external id.
+    pub fn get_by_external_id(&self, external_id: &str) -> Option<&EntityMetadata> {
+        let mqtt_id = self.external_id_index.get(external_id)?;
+        self.get(mqtt_id)
+    }
+
+    /// Returns the MQTT identifier of the main device.
+    ///
+    /// The main device is an entity with `@type: "device"`.
+    pub fn main_device(&self) -> EntityIdRef {
+        self.main_device.as_str()
+    }
+
+    /// Returns MQTT identifiers of child devices of a given device.
+    pub fn child_devices(&self, entity_topic: EntityIdRef) -> Vec<EntityIdRef> {
+        self.entities
+            .iter()
+            .filter(|(_, e)| {
+                e.parent.as_ref().is_some_and(|p| p == entity_topic) && e.r#type == "child-device"
+            })
+            .map(|(k, _)| k.as_str())
+            .collect()
+    }
+
+    /// Returns MQTT identifiers of services running on a given device.
+    pub fn services(&self, entity_topic: EntityIdRef) -> Vec<EntityIdRef> {
+        self.entities
+            .iter()
+            .filter(|(_, e)| {
+                e.parent.as_ref().is_some_and(|p| p == entity_topic) && e.r#type == "service"
+            })
+            .map(|(k, _)| k.as_str())
+            .collect()
+    }
+
+    /// Updates entity store state based on the content of the entity
+    /// registration message.
+    ///
+    /// It can register a new entity in the store or update already registered
+    /// entity, returning a list of all entities affected by the update, e.g.:
+    ///
+    /// - when adding/removing a child device or service, the parent is affected
+    pub fn update(&mut self, message: EntityRegistrationMessage) -> Result<Vec<EntityId>, Error> {
+        if message.r#type == "device" && message.mqtt_id != self.main_device {
+            return Err(Error::MainDeviceAlreadyRegistered(
+                self.main_device.as_str().into(),
+            ));
+        }
+
+        let mut affected_entities = vec![];
+
+        let parent = match message.r#type.as_str() {
+            "child-device" => message.parent.or(Some(self.main_device.clone())),
+            "service" => message.parent.or(Some(self.main_device.clone())),
+            _ => None,
+        };
+
+        // parent device is affected if new device is its child
+        if let Some(parent) = &parent {
+            if !self.entities.contains_key(parent) {
+                return Err(Error::NoParent(parent.clone().into_boxed_str()));
+            }
+
+            affected_entities.push(parent.clone());
+        }
+
+        let external_id = self.external_id(&message.mqtt_id);
```

Comment:
If the message provides an `external_id` this one must be used.

```suggestion
        let external_id = message.external_id.unwrap_or_else(||self.external_id(&message.mqtt_id));
```